samples/ChatSample/DefaultPresenceManager.cs

```diff
@@ -0,0 +1,84 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.SignalR;
+using Microsoft.AspNetCore.Sockets;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+
+namespace ChatSample
+{
+    public class DefaultPresenceManager<THub> : IPresenceManager where THub : HubWithPresence
+    {
+        private IHubContext<THub> _hubContext;
+        private HubLifetimeManager<THub> _lifetimeManager;
+        private readonly IServiceScopeFactory _serviceScopeFactory;
+        private readonly ILogger _logger;
+
+        public DefaultPresenceManager(IHubContext<THub> hubContext, HubLifetimeManager<THub> lifetimeManager,
+            IServiceScopeFactory serviceScopeFactory, ILoggerFactory loggerFactory)
+        {
+            _hubContext = hubContext;
+            _lifetimeManager = lifetimeManager;
+            _serviceScopeFactory = serviceScopeFactory;
+            _logger = loggerFactory.CreateLogger<DefaultPresenceManager<THub>>();
+        }
+
+        private readonly ConcurrentDictionary<Connection, UserDetails> usersOnline
+            = new ConcurrentDictionary<Connection, UserDetails>();
+
+        public Task<IEnumerable<UserDetails>> UsersOnline()
+            => Task.FromResult(usersOnline.Values.AsEnumerable());
+
+        public async Task UserJoined(Connection connection)
+        {
+            var user = new UserDetails(connection.ConnectionId, connection.User.Identity.Name);
+
+            await Notify(hub => hub.OnUserJoined(user));
+
+            usersOnline.TryAdd(connection, user);
+        }
+
+        public async Task UserLeft(Connection connection)
+        {
+            if (usersOnline.TryRemove(connection, out UserDetails user))
+            {
+                await Notify(hub => hub.OnUserLeft(user));
+            }
+        }
+
+        private async Task Notify(Func<THub, Task> invocation)
+        {
+            foreach (var connection in usersOnline.Keys)
+            {
+                using (var scope = _serviceScopeFactory.CreateScope())
```

Comment:
This is the grossest part of this change. Creating a new hub instance here feels kinda wrong as it requires mimicking the HubEndPoint logic... At least it's a sample and we get to see what's required to make it work.